### PR TITLE
Create tests for components defined as refs twice

### LINKED ISSUE #71: Test and document what happens when an element is defined as `ref twice
Most "common" use case, is when an element needs controlling from its parent and grandparent (mostly when dealing with wrappers).

Need to test:
* if the component is only initialised once
* if both refs share the same component instance
* what happens when using bindings from both places that DON'T overlap
* what happens when using bindings from both places that DO overlap

And more advanced:
* what happens with component collections, that maybe partly overlap between parents and children

### PR DESCRIPTION
**Pull request by jspolancor**
With this tests we're trying to solve some questions about defining child components as 'refs' in multiple parents.

@ThaNarie This is what I discovered so far and the questions I have

_test if the component is only initialised once:_ I'm testing the mounting of the child component, it gets mounted once for each parent, is testing the mounting good enough?

_test if both refs share the same component instance:_ I'm unsure how to test this, I could get the global muban instance to test if the child ref exists only once in the instance, does that make sense? the other solution is from the grandparent component look for the parent's child instance like, grandparent.__instance.refs.parent.__instance.refs.child.__instance.uid but feels hacky to me

_what happens when using bindings from both places that DON'T overlap:_
_what happens when using bindings from both places that DO overlap:_
_what happens with component collections, that maybe partly overlap between parents and children:_

The answer for this questions is the same, the bindings are applied from the child like Child > Parent > Grandparent, the grandparent bindings will always overwrite the parent's if they overlap, if they don't overlap then the child updates made by the parent bindings will be preserved, same rule applies to collections

After finishing the tests i still need to write the docs, do you think addding a new section to the [component bindings](https://mubanjs.github.io/muban/api/bindings.html#component-bindings) docs is good enough 🍻

### CODE REVIEW

**Comment by ThaNarie** [comment]
> **test if the component is only initialised once**
> I'm testing the mounting of the child component, it gets mounted once for each parent, is testing the mounting good enough?

Yes it's good enough for the test, but I don't think this is what we want.
It should only be mounted once, and applied bindings should be set on the same mounted instance.
It'll probably not make a difference in the observed behaviour (which seems to be working mostly okay, luckily), but it does work it shouldn't.

So you could create a separate ticket to change that behaviour.

> **test if both refs share the same component instance**
> I'm unsure how to test this...

As you hinted, the only real way to get to internal information is through the `__instance`. Seems like step 1 is getting access to both Grandparent and Parent, and from there get an instance to the ref for each and compare that (perhaps on different level, the object reference, a uuid, the component instance on the ref, etc).

If we don't have enough/easy internal information to check these kind of things, we might need to invest in things to improve this, like more/easier internal access, and/or utils to work with them, etc.

Let me know if this is enough to get you going, or if I should start looking into it a bit more :)

> **what happens when using bindings from both places that DON'T overlap:
> what happens when using bindings from both places that DO overlap:
> what happens with component collections, that maybe partly overlap between parents and children:**
> The answer for this questions is the same...

Nice, although it surprises me that the `parentAttr` is `null`, while the grandparent doesn't override this one. Strangely the `isActive` does seem to work. Seems like a bug?

Also, would be good to test some compound bindings, like adding different `$element` bindings, and within that, different `attr` and `event` bindings that get merged. So far you've only tested individual component props.


> After finishing the tests i still need to write the docs, do you think addding a new section to the component bindings docs is good enough?

I think this is more of a chapter for in the "Guide" area, and then mention and link it from the `ignoreGuard` option in the API docs.

File: src/lib/Component.test.ts [hunk 1]
```diff
@@ -0,0 +1,124 @@
...
+
+      // The component should be mounted two times, one for Parent one for GrandParent
+      expect(mountChild).toHaveBeenCalledTimes(2);
+      // The click function from the parent gets called first, then the grapndparent's
+      expect(handleChildClickFromParentOrder).toBeLessThan(handleChildClickFromGrandParentOrder);
+      expect(handleChildClickFromGrandParent).toHaveBeenCalledTimes(1);
+      expect(handleChildClickFromParent).toHaveBeenCalledTimes(1);
+      // bindings are applied from the child component like Child > Parent > Grandparent
+      // The final textContent should be the one defined in the grandparent
+      expect(child.textContent).toBe('text from grandparent');
```
Comment (ThaNarie) [inline src/lib/Component.test.ts:116]: This could make use of the DOM matchers :)

**Comment by jspolancor** [comment]
> It should only be mounted once, so you could create a separate ticket to change that behaviour.

issue [here](https://github.com/mubanjs/muban/issues/98)

> If we don't have enough/easy internal information to check these kind of things, we might need to invest in things to improve this, like more/easier internal access, and/or utils to work with them, etc.

I created [this issue](https://github.com/mubanjs/muban/issues/97) for ref component instance extraction

> Nice, although it surprises me that the `parentAttr` is `null`, while the grandparent doesn't override this one. Strangely the `isActive` does seem to work. Seems like a bug?
I don't think it's a bug, well, it might be, something related to the double mounting, the child component gets remounted and that particular prop is not defined in the grandparent, the prop becomes falsy and the attr gets removed

> Also, would be good to test some compound bindings, like adding different `$element` bindings, and within that, different `attr` and `event` bindings that get merged. So far you've only tested individual component props.

A few tests were added

> I think this is more of a chapter for in the "Guide" area, and then mention and link it from the `ignoreGuard` option in the API docs.

I think it's better to finish the two new issues to be completely sure of what's going on before updating the docs

File: src/lib/Component.test.ts (hunk 1 above)
Comment (jspolancor) [reply src/lib/Component.test.ts:116]: I had a hard time trying to install and configure jest-dom, could you help me with that please?

Comment (ThaNarie) [reply src/lib/Component.test.ts:116]: I pushed a commit in your branch to make this work :)

**Comment by ThaNarie** [comment]
We decided to not allow double refs on a single element, which has been enforced in #99

**Pull request closed without merging**